File: src/main/java/app/tweditor/LocalizedString.java
package app.tweditor;

import java.util.ArrayList;
import java.util.List;
import java.util.ListIterator;

public class LocalizedString extends DBElementValue
  implements Cloneable
{
  private int stringReference;
  private List<LocalizedSubstring> substringList;

  public LocalizedString(int reference)
  {
    this.stringReference = reference;
    this.substringList = new ArrayList<>(4);
  }

  public void addSubstring(LocalizedSubstring substring)
  {
    int language = substring.getLanguage();
    int gender = substring.getGender();
    ListIterator<LocalizedSubstring> li = this.substringList.listIterator();
    boolean found = false;
    while (li.hasNext()) {
      LocalizedSubstring oldSubstring = li.next();
      if ((oldSubstring.getLanguage() == language) && (oldSubstring.getGender() == gender)) {
        li.set(substring);
        found = true;
        break;
      }
    }

    if (!found)
      this.substringList.add(substring);
  }

  public int getStringReference()
  {
    return this.stringReference;
  }

  public void setStringReference(int reference)
  {
    this.stringReference = reference;
  }

  public int getSubstringCount()
  {
    return this.substringList.size();
  }

  public LocalizedSubstring getSubstring(int index)
  {
    return this.substringList.get(index);
  }

  public List<LocalizedSubstring> getSubstrings()
  {
    return this.substringList;
  }

  public void setSubstring(int index, LocalizedSubstring substring)
  {
    this.substringList.set(index, substring);
  }

  public LocalizedSubstring getSubstring(int language, int gender)
  {
    LocalizedSubstring value = null;
    for (LocalizedSubstring substring : this.substringList) {
      if ((substring.getLanguage() == language) && (substring.getGender() == gender)) {
        value = substring;
        break;
      }
    }

    return value;
  }

  public LocalizedString clone()
  {
    LocalizedString clonedString = (LocalizedString)super.clone();

    clonedString.substringList = new ArrayList<>(this.substringList.size());
    for (LocalizedSubstring substring : this.substringList) {
      clonedString.substringList.add(substring.clone());
    }
    return clonedString;
  }
}

File: src/main/kotlin/app/tweditor/LocalizedString.kt
package app.tweditor

class LocalizedString(stringReference: Int) : DBElementValue() {
    var stringReference: Int = stringReference
    private var substringList: MutableList<LocalizedSubstring> = ArrayList(4)

    fun addSubstring(substring: LocalizedSubstring) {
        val existing = substringList.indexOfFirst { candidate ->
            candidate.language == substring.language && candidate.gender == substring.gender
        }
        if (existing >= 0) {
            substringList[existing] = substring
        } else {
            substringList.add(substring)
        }
    }

    fun getSubstringCount(): Int = substringList.size

    fun getSubstring(index: Int): LocalizedSubstring = substringList[index]

    fun getSubstrings(): List<LocalizedSubstring> = substringList

    fun getSubstring(language: Int, gender: Int): LocalizedSubstring? {
        return substringList.find { candidate ->
            candidate.language == language && candidate.gender == gender
        }
    }

    public override fun clone(): LocalizedString {
        val cloned = super.clone() as LocalizedString
        cloned.substringList = ArrayList(substringList.size)
        for (substring in substringList) {
            cloned.substringList.add(substring.clone())
        }
        return cloned
    }
}
